core/src/main/scala/sttp/client/testing/SttpBackendStub.scala

```diff
@@ -211,3 +282,100 @@ object SttpBackendStub {
     }
   }
 }
+
+/**
+  * Websockets use can differ, for akka backend crucial part is the handler.
+  * For other backends
+  */
+private[testing] sealed trait WhenOpenWebsocket[F[_], +WS_HANDLER[_]]
+private[testing] case class UseHandler[F[_], WS_RESULT, WS_HANDLER[_]](
+    headers: Headers,
+    useHandler: Any => WS_RESULT
+) extends WhenOpenWebsocket[F, WS_HANDLER]
+
+private[testing] case class ReturnWebsocketResponse[F[_], WS_RESULT](headers: Headers, response: F[WS_RESULT])
+    extends WhenOpenWebsocket[F, NothingT]
+
+/**
+  * Simple stub for web sockets that uses queue of events for receive.
+  */
+class WebSocketStub[F[_], S](
+    events: mutable.Queue[F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]],
+    initialState: S,
+    makeNewEvents: (S, WebSocketFrame) => (S, List[F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]]])
+)(implicit
+    val monad: MonadError[F]
+) extends WebSocket[F] {
+
+  private var state: S = initialState
+
+  override def isOpen: F[Boolean] = monad.unit(true)
+
+  override def receive: F[Either[WebSocketEvent.Close, WebSocketFrame.Incoming]] =
+    events.dequeueFirst(_ => true) match {
+      case Some(event) => event
+      case None        => monad.error(new Exception("Unexpected 'receive', no more prepared events."))
+    }
+
+  override def send(frame: WebSocketFrame, isContinuation: Boolean): F[Unit] =
+    monad.eval {
+      val (newState, newEvents) = makeNewEvents(state, frame)
+      newEvents.foreach(events.enqueue(_))
+      state = newState
+    }
+
+  /** Returns stub that has same prepared messages but adds new messages to receive in reaction to send.
+    *
+    * @param addReceived
+    */
+  def respondUsing(addReceived: WebSocketFrame => List[WebSocketFrame.Incoming]): WebSocketStub[F, Unit] =
+    respondUsingF(
+      addReceived.andThen(_.map(m => monad.unit(Right(m): Either[WebSocketEvent.Close, WebSocketFrame.Incoming])))
+    )
+
+  /** Returns stub that has same prepared messages but adds new messages to receive in reaction to send.
+    *
+    * @param addReceived
+    */
+  def respondUsingF(
```

Comment:
it's also probably important to mention (in the docs) that this *replaces* the current event-producing logic. (And the enqueued events remain as-is)